pkg/packetfilter/nftables/rule_conversion.go

```diff
@@ -0,0 +1,229 @@
+		ruleSpec = append(ruleSpec, "ip", "protocol", "tcp")
+		if dPort != "" {
+			ruleSpec = append(ruleSpec, "tcp", "dport", dPort)
+		}
+	case packetfilter.RuleProtoICMP:
+		ruleSpec = append(ruleSpec, "ip", "protocol", "icmp")
+	case packetfilter.RuleProtoAll:
+	case packetfilter.RuleProtoUndefined:
+	}
+
+	return ruleSpec
+}
+
+func mssClampToRuleSpec(ruleSpec []string, clampType packetfilter.MssClampType, mssValue string) []string {
+	switch clampType {
+	case packetfilter.UndefinedMSS:
+	case packetfilter.ToPMTU:
+		ruleSpec = append(ruleSpec, "size", "set", "rt", "mtu")
+	case packetfilter.ToValue:
+		ruleSpec = append(ruleSpec, "size", "set", mssValue)
+	}
+
+	return ruleSpec
+}
+
+func setToRuleSpec(ruleSpec []string, srcSetName, destSetName string) []string {
+	if srcSetName != "" {
+		ruleSpec = append(ruleSpec, "ip", "saddr", "@"+srcSetName)
+	}
+
+	if destSetName != "" {
+		ruleSpec = append(ruleSpec, "ip", "daddr", "@"+destSetName)
+	}
+
+	return ruleSpec
+}
+
+func toNftRuleSpec(rule *packetfilter.Rule) string {
+	ruleSpec := protoToRuleSpec([]string{}, rule.Proto, rule.DPort)
+
+	if rule.SrcCIDR != "" {
+		ruleSpec = append(ruleSpec, "ip", "saddr", rule.SrcCIDR)
+	}
+
+	if rule.DestCIDR != "" {
+		ruleSpec = append(ruleSpec, "ip", "daddr", rule.DestCIDR)
+	}
+
+	if rule.MarkValue != "" && rule.Action != packetfilter.RuleActionMark {
+		// syntax for MarkValue '0xc0000': meta mark & 0xc0000 == 0xc0000
+		ruleSpec = append(ruleSpec, "meta", "mark", "&", rule.MarkValue, "==", rule.MarkValue)
+	}
+
+	ruleSpec = setToRuleSpec(ruleSpec, rule.SrcSetName, rule.DestSetName)
+
+	if rule.OutInterface != "" {
+		ruleSpec = append(ruleSpec, "oifname", rule.OutInterface)
+	}
+
+	if rule.InInterface != "" {
+		ruleSpec = append(ruleSpec, "iifname", rule.InInterface)
+	}
+
+	if rule.Action == packetfilter.RuleActionMss {
+		ruleSpec = append(ruleSpec, "tcp", "flags", "syn / syn,rst")
+	}
+
+	ruleSpec = append(ruleSpec, "counter")
+	ruleSpec = append(ruleSpec, ruleActionToStr[rule.Action]...)
+
+	if rule.Action == packetfilter.RuleActionJump {
+		ruleSpec = append(ruleSpec, rule.TargetChain)
+	}
+
+	if rule.SnatCIDR != "" {
+		ruleSpec = append(ruleSpec, "to", rule.SnatCIDR)
+	}
+
+	if rule.DnatCIDR != "" {
+		ruleSpec = append(ruleSpec, "to", rule.DnatCIDR)
+	}
+
+	ruleSpec = mssClampToRuleSpec(ruleSpec, rule.ClampType, rule.MssValue)
+
+	if rule.MarkValue != "" && rule.Action == packetfilter.RuleActionMark {
+		// syntax for MarkValue '0xc0000': set mark | 0xc0000
+		ruleSpec = append(ruleSpec, "set", "mark", "|", rule.MarkValue)
+	}
+
+	str := strings.Join(ruleSpec, " ")
+
+	logger.V(log.TRACE).Infof("toNftRuleSpec: from %q to %q", rule, str)
+
+	return str
+}
+
+func mustWriteString(buf *bytes.Buffer, s string) {
+	_, err := buf.WriteString(s + " ")
+	utilruntime.Must(err)
+}
+
+func mustWriteUint32(buf *bytes.Buffer, i uint32) {
+	mustWriteString(buf, strconv.FormatInt(int64(i), 10))
+}
+
+func SerializeRule(rule *packetfilter.Rule) string {
```

Comment:
Considering the proposed Serialize/Deserialize approach and the potential need to address future backward compatibility, I wonder if it might be simpler only to store compressed ruleSpec string in Comment field. This way, decoding would consistently rely on nft syntax and no need to address potential backward compatibility in future/

Reply:
You mean use gzip compression?  I played around with that and it ends up being larger in size due to added overhead. Compression really only yields benefit for data sizes much larger than what we're dealing with here. The nft syntax is pretty "wordy" - I don't know what way we could reduce the size more efficiently than the simple serialization approach I implemented. It's probably unlikely we would change the Rule structure in the future in any significant way, if at all,  but, even so,  it should easy to handle backwards compatibility.

Reply:
Yep, I meant gzip compression.
Also not sure how likely it would be to change packet filter rule in future, maybe we should at least leave a note that Serialize/Deserialize should be updated if the pf rule changed.